src/renderers/WebGLRenderer.js

```diff
@@ -1236,36 +1247,82 @@ THREE.WebGLRenderer = function ( parameters ) {
 
 	function pushImmediateRenderItem( object ) {
 
+		var array, index;
+
+		// allocate the next position in the appropriate array
+
 		if ( object.material.transparent ) {
 
-			transparentImmediateObjects.push( object );
+			array = transparentImmediateObjects;
+			index = ++ transparentImmediateObjectsLastIndex;
+
+		} else {
+
+			array = opaqueImmediateObjects;
+			index = ++ opaqueImmediateObjectsLastIndex;
+
+		}
+
+		// recycle existing position or grow the array
+
+		if ( index < array.length ) {
+
+			array[ index ] = object;
 
 		} else {
 
-			opaqueImmediateObjects.push( object );
+			// assert( index === array.length );
+			array.push( object );
 
 		}
 
+
 	}
 
 	function pushRenderItem( object, geometry, material, z, group ) {
 
-		var renderItem = {
-			id: object.id,
-			object: object,
-			geometry: geometry,
-			material: material,
-			z: _vector3.z,
-			group: group
-		};
+		var array, index;
+
+		// allocate the next position in the appropriate array
 
 		if ( material.transparent ) {
 
-			transparentObjects.push( renderItem );
+			array = transparentObjects;
+			index = ++ transparentObjectsLastIndex;
 
 		} else {
 
-			opaqueObjects.push( renderItem );
+			array = opaqueObjects;
+			index = ++ opaqueObjectsLastIndex;
+
+		}
+
+		// recycle existing render item or grow the array
+
+		var renderItem = array[ index ];
+
+		if ( renderItem !== undefined ) {
```

Comment:
The way I see it, the code is just very explicit. As said, I'm not at all sure it's really needed, but different code would not explicitly state what is supposed to happen and instead rely on optimizations to fuse JavaScript statements. One could even argue that it's cleaner for this reason.

I agree that the redundancy is not all too pretty, but the difference in GC load gives enough of a reason to better be safe than sorry. Both code paths are right next to each other in the same function (that's why I didn't factorize) and the overall maintainability should be very well within acceptable bounds.

Oh BTW, Knuth's famous "premature optimization" statement reads in its full beauty:

> We should forget about small efficiencies, say about 97% of the time: premature optimization is the root of all evil. *Yet we should not pass up our opportunities in that critical 3%.* 

...and we're looking at the only spot in the renderer that uses dynamic memory every frame.

Hope this discussion makes you sleep better after merging my obscene code :-D.